std_execution.bs

```diff
@@ -1247,6 +1248,133 @@ talk about the semantics of some of the algorithms proposed in this paper.
+
+2.  Detach from the computation - let the operation continue in the background -
+    like an implicit call to `std::thread::detach()`. While this approach can
+    work in some circumstances for some kinds of applications, in general it is
+    also pretty user-hostile; it makes it difficult to reason about the safe
+    destruction of resources used by these eager operations. In general, detached
+    work necessitates some kind of garbage collection; <i>e.g.</i>,
+    `std::shared_ptr`, to ensure resources are kept alive until the operations
+    complete, and can make clean shutdown nigh impossible.
+
+3.  Block in the destructor until the operation completes. This approach is
+    probably the safest to use as it preserves the structured nature of the
+    concurrent operations, but also introduces the potential for deadlocking the
+    application if the completion of the operation depends on the current thread
+    making forward progress.
+
+      The risk of deadlock might occur, for example, if a thread-pool with a
+    small number of threads is executing code that creates a sender representing
+    an eagerly-executing operation and then calls the destructor of that sender
+    without joining it (e.g. because an exception was thrown). If the current
+    thread blocks waiting for that eager operation to complete and that eager
+    operation cannot complete until some entry enqueued to the thread-pool's
+    queue of work is run then the thread may wait for an indefinite amount of
+    time. If all thread of the thread-pool are simultaneously performing such
+    blocking operations then deadlock can result.
+
+There are also minor variations on each of these choices. For example:
+
+4.  A variation of (1): Call `std::terminate` if an eager sender is destructed
+    without joining it. This is the approach that `std::thread` destructor
+    takes.
+
+5.  A variation of (2): Request cancellation of the operation before detaching.
+    This reduces the chances of operations continuing to run indefinitely in the
+    background once they have been detached but does not solve the
+    lifetime- or shutdown-related challenges.
+
+6.  A variation of (3): Request cancellation of the operation before blocking on
+    its completion. This is the strategy that `std::jthread` uses for its
+    destructor. It reduces the risk of deadlock but does not eliminate it.
+
+### Eager senders complicate algorithm implementations ### {#design-lazy-algorithms-complexity}
+
+Algorithms that can assume they are operating on senders with strictly lazy
+semantics are able to make certain optimizations that are not available if
+senders can be potentially eager. With lazy senders, an algorithm can safely
+assume that a call to `execution::start` on an operation state strictly happens
+before any receiver methods are called. This frees the algorithm from needing to
+resolve potential race conditions between the start and completion of the work.
```

Comment:
The guarantee that completion-signalling methods are not called on the receiver until after `start()` is called is guaranteed for all senders, so I don't think that's a distinguishing feature of eager vs lazy.

Instead, the main difference is whether or not the logic of a particular algorithm is guaranteed not to run until after the sender has been connected and started.

e.g. we might write something like:
```
sender auto example(sender src) {
  return sequence(do_a(), then(src, [] { do_b(); }), do_c());
}
```
An algorithm like this might expect that do_a(), do_b() and do_c() are sequenced and therefore do not need synchronisation. If the `then()` algorithm might actually execute the lambda eagerly then this might break the algorithm's assumption as it could be possible for `do_b()` to run before `do_a()` has completed.